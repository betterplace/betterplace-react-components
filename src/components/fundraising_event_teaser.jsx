import React from 'react'
import PropTypes from 'prop-types'
import { formatAmount } from '../helpers/format_amount'

export const FundraisingEventTeaser = (props) => {
  const fundraisingEvent = props.fundraisingEvent
  const locale = props.locale || document.documentElement.lang || 'de'
  const href = props.href || fundraisingEvent.links.find(link => link.rel === 'platform')?.href
  const fundraisingEventImageUrl = fundraisingEvent.profile_picture.links.find(link => link.rel === 'fill_410x214')?.href
  const donationsAmount = formatAmount({ cents: fundraisingEvent.donated_amount_in_cents, locale: locale })
  const contactImageUrl = fundraisingEvent.contact.picture.links.find(link => link.rel === 'fill_100x100')?.href
  const handleClick = (event) => { props.onClick && props.onClick(event, fundraisingEvent) }

  return <a
    title={fundraisingEvent.title}
    onClick={handleClick}
    className='donatable-teaser'
    href={href}
    target={props.openInTab ? '_blank' : '_parent'}
  >
    {props.children /* mount point, e.g. for injecting VisibilitySensor */}

    <div className='donatable-teaser--profile-picture' style={{backgroundImage: `url(${fundraisingEventImageUrl})`}} alt={fundraisingEvent.title} />

    <h2 className='donatable-teaser--title' dangerouslySetInnerHTML={{__html: fundraisingEvent.title}} />

    <div className='donatable-teaser--divider'>
      <div className='donatable-teaser--factlist'>
        <div><span dangerouslySetInnerHTML={{ __html: fundraisingEvent.contact.name }}/></div>
        <div><span>{fundraisingEvent.donations_count}</span> {fundraisingEvent.donations_count === 1 ? (locale === 'de' ? 'Spende' : 'donation') : (locale === 'de' ? 'Spenden' : 'donations')}</div>
        <div><span>{donationsAmount}</span> {locale === 'de' ? 'gesammelt' : 'collected'}</div>
      </div>
      <img style={{ borderRadius: '50%', border: '1px solid #cccccc'}} src={contactImageUrl} alt={fundraisingEvent.contact.name} className='donatable-teaser--carrier-logo' />
    </div>
  </a>
}

FundraisingEventTeaser.propTypes = {
  font:             PropTypes.string,
  href:             PropTypes.string,
  onClick:          PropTypes.func,
  fundraisingEvent: PropTypes.object.isRequired,
}
