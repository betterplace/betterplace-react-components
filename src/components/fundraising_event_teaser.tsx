import { components } from '../apiV4'
import React from 'react'

import { formatAmount } from '../helpers/format_amount'

export type FundraisingEvent = components['schemas']['FundraisingEventResult'] | components['schemas']['SearchResult']
export type Links = components['schemas']['FundraisingEventResult']['links'] &
  components['schemas']['SearchResult']['links']
export type FundraisingEventTeaserProps = {
  locale?: string
  href?: string
  font?: string
  openInTab?: boolean
  fundraisingEvent: FundraisingEvent
  onClick?: (event: React.MouseEvent, fundraisingEvent: FundraisingEvent) => void
}

const pluralizeDonation = (locale: string, count: number) => {
  if (count === 1) return locale === 'de' ? 'Spende' : 'donation'
  return locale === 'de' ? 'Spenden' : 'donations'
}

export const FundraisingEventTeaser: React.FC<FundraisingEventTeaserProps> = ({
  fundraisingEvent,
  locale = document.documentElement.lang || 'de',
  href: href_,
  onClick,
  openInTab,
  children,
}) => {
  const href = href_ || (fundraisingEvent.links as Links).find((link) => link.rel === 'platform')?.href
  const fundraisingEventImageUrl = fundraisingEvent.profile_picture?.links.find(
    (link) => link.rel === 'fill_410x214'
  )?.href
  const donationsAmount = formatAmount({ cents: fundraisingEvent.donated_amount_in_cents, locale: locale })
  const contactImageUrl = fundraisingEvent.contact?.picture?.links.find((link) => link.rel === 'fill_100x100')?.href
  const handleClick = (event: React.MouseEvent) => {
    onClick && onClick(event, fundraisingEvent)
  }

  const donationsTxt = pluralizeDonation(locale, fundraisingEvent.donations_count)

  return (
    <a
      title={fundraisingEvent.title}
      onClick={handleClick}
      className="donatable-teaser"
      href={href}
      target={openInTab ? '_blank' : '_parent'}
      rel="noreferrer"
    >
      {children /* mount point, e.g. for injecting VisibilitySensor */}

      <div
        className="donatable-teaser--profile-picture"
        style={{ backgroundImage: `url(${fundraisingEventImageUrl})` }}
        title={fundraisingEvent.title}
      />

      <h2 className="donatable-teaser--title" dangerouslySetInnerHTML={{ __html: fundraisingEvent.title }} />

      <div className="donatable-teaser--divider">
        <div className="donatable-teaser--factlist">
          <div>
            <span dangerouslySetInnerHTML={{ __html: fundraisingEvent.contact?.name ?? '' }} />
          </div>
          <div>
            <span>{fundraisingEvent.donations_count}</span> {donationsTxt}
          </div>
          <div>
            <span>{donationsAmount}</span> {locale === 'de' ? 'gesammelt' : 'collected'}
          </div>
        </div>
        <img
          style={{ borderRadius: '50%', border: '1px solid #cccccc' }}
          src={contactImageUrl}
          alt={fundraisingEvent.contact?.name ?? ''}
          className="donatable-teaser--carrier-logo"
        />
      </div>
    </a>
  )
}
