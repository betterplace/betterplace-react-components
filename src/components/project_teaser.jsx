import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { formatAmount } from '../helpers/format_amount'

export class ProjectTeaser extends React.Component {
  handleClick = (event) => {
    this.props.onClick && this.props.onClick(event, this.props.project)
  }

  get href() {
    return this.props.href || this.props.project.links.find(link => link.rel == 'platform').href
  }

  get projectImageUrl() {
    const link = this.props.project.profile_picture.links.find(link => link.rel == 'fill_410x214')
    return link && link.href
  }

  get orgaImageUrl() {
    const link = this.props.project.carrier.picture.links.find(link => link.rel == 'fill_100x100')
    return link && link.href
  }

  render() {
    const project = this.props.project
    const locale = this.props.locale || document.documentElement.lang || 'de'
    const openAmount = formatAmount({cents: this.props.project.open_amount_in_cents, locale: locale})
    const donationsCount = new Intl.NumberFormat(locale).format(project.donations_count)
    const openAmountCaption = locale == 'de' ? 'fehlen noch' : 'still needed'
    const donationsCountCaption = locale == 'de' ? 'Spenden' : 'donations'

    return (
      <Wrapper
        title={project.title}
        onClick={this.handleClick}
        className={this.props.className}
        href={this.href}
        target='_parent'
        font={this.props.font}
        textColor={this.props.textColor}
        showDescription={this.props.showDescription}
      >
        { this.props.children /* mount point, e.g. for injecting VisibilitySensor */ }

        <ProfilePicture src={ this.projectImageUrl } alt={ project.title } />

        <Title dangerouslySetInnerHTML={ { __html: project.title } } />
        { this.props.showDescription && <Description dangerouslySetInnerHTML={ { __html: project.description } } /> }

        <Divider>
         <FactList>
           <div>{project.city && <span>{ project.city }, </span>}{project.country}</div>
           <div><span>{ donationsCount }</span> {donationsCountCaption}</div>
           <div><span>{ openAmount }</span> {openAmountCaption}</div>
         </FactList>
         <CarrierLogo src={ this.orgaImageUrl } alt={ project.carrier.name } />
        </Divider>

        <Progress color={ this.props.progressbarColor || '#bad304' } value={ project.progress_percentage }>
          <div className='bar'></div>
        </Progress>
      </Wrapper>
    )
  }
}

ProjectTeaser.propTypes = {
  className:        PropTypes.string,
  font:             PropTypes.string,
  textColor:        PropTypes.string,
  href:             PropTypes.string,
  onClick:          PropTypes.func,
  progressbarColor: PropTypes.string,
  project:          PropTypes.object.isRequired,
  showDescription:  PropTypes.bool
}

const Wrapper = styled.a`
  cursor: pointer;
  font-family: ${ props => props.font ? props.font : '"Fira Sans", Arial, sans-serif' };
  border: 1px solid #ccc;
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  color: ${ props => props.textColor ? props.textColor : '#000000' };
  font-weight: 300;
  height: ${ props => props.showDescription ? '406px' : '319px' };
  box-sizing: border-box;
  transition: 0.15s box-shadow ease-out;

  @media (max-width: 767px) {
    width: 100%;
  }

  &:hover,
  &:focus {
    text-decoration: none;
    box-shadow: 0 0 10px 0 rgba(0,0,0,0.5);
    color: inherit;

    h2 {
      text-decoration: underline;
    }
  }
`

const Title = styled.h2`
  display: block;
  margin: 10px;
  font-size: 16px;
  line-height: 22px;
  height: 41px;
  font-weight: bold;
  overflow: hidden;
`

const Description = styled.div`
  position: relative;
  height: 71px;
  overflow: hidden;
  padding: 0 13px;
  font-size: 14px;
  line-height: 18px;
  color: #6a6a6a;

  &:after {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    height: 18px;
    background: linear-gradient(180deg, rgba(255,255,255,0) 0%, rgba(255,255,255,1) 100%);
  }
`

const Divider = styled.div`
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  flex-grow: 1;
  margin-bottom: 10px;
`

const FactList = styled.div`
  font-size: 14px;
  line-height: 22px;
  padding-left: 13px;
  width: calc(100% - 100px);

  > div {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    > span {
      font-weight: bold;
    }
  }
`

const CarrierLogo = styled.img`
  width: 65px;
  height: 65px;
  margin-right: 13px;
  overflow: hidden;
`

const Progress = styled.div`
  background-color: #eee;
  height: 10px;
  width: 100%;
  border-bottom-left-radius: 3px;
  border-bottom-right-radius: 3px;
  overflow: hidden;

  > div {
    width: ${props => props.value}%;
    background: ${ props => props.color || 'red' };
    transition: width 0.2s linear;
    height: 100%;
  }
`

const ProfilePicture = styled.div`
	background: url(${props => props.src});
  height: 150px;
  background-size: cover;
`;
