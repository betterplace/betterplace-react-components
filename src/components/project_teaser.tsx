import { components } from 'apiV4'
import PropTypes from 'prop-types'
import React from 'react'

import { formatAmount } from '../helpers/format_amount'
export type Project = components['schemas']['ProjectResult']
export type ProjectTeaserProps = {
  locale?: string
  project: Project
  onClick?: (event: React.MouseEvent, project: Project) => void
  showDescription?: boolean
  bottomContent?: React.ReactNode
  href?: string
  showCarrierName?: boolean
  openInTab?: boolean
  className?: string
  font?: string
  textColor?: string
  progressbarColor?: string
  progressbarBackgroundColor?: string
  extraHeight?: number
}
export const ProjectTeaser: React.FC<ProjectTeaserProps> = (props) => {
  const project = props.project
  const locale = props.locale || document.documentElement.lang || 'de'
  const openAmount = formatAmount({ cents: project.open_amount_in_cents, locale: locale })
  const donationsCount = new Intl.NumberFormat(locale).format(project.donations_count)
  const openAmountCaption = locale === 'de' ? 'fehlen noch' : 'still needed'
  const donationsCountCaption = locale === 'de' ? 'Spenden' : 'donations'
  const href = props.href || props.project.links.find((link) => link.rel === 'platform')?.href
  const projectImageUrl = props.project.profile_picture?.links.find((link) => link.rel === 'fill_410x214')?.href
  const orgaImageUrl = props.project.carrier?.picture?.links.find((link) => link.rel === 'fill_100x100')?.href
  const handleClick = (event: React.MouseEvent) => {
    props.onClick && props.onClick(event, project)
  }

  return (
    <a
      title={project.title}
      onClick={handleClick}
      className={`donatable-teaser ${props.className ?? ''} ${
        props.showDescription ? 'donatable-teaser--with-description' : ''
      } ${props.bottomContent ? 'donatable-teaser--with-bottom-content' : ''}`}
      href={href}
      target={props.openInTab ? '_blank' : '_parent'}
      rel="noreferrer"
    >
      {props.children /* mount point, e.g. for injecting VisibilitySensor */}

      <div
        className="donatable-teaser--profile-picture"
        style={{ backgroundImage: `url(${projectImageUrl})` }}
        title={project.title}
      />

      <h2 className="donatable-teaser--title" dangerouslySetInnerHTML={{ __html: project.title }} />

      {props.showDescription && (
        <div className="donatable-teaser--description" dangerouslySetInnerHTML={{ __html: project.description }} />
      )}

      <div className="donatable-teaser--divider">
        <div className="donatable-teaser--factlist">
          <div>
            {project.city && <span>{project.city}</span>}, {project.country}
          </div>
          <div>
            {props.showCarrierName ? (
              <span>{props.project.carrier?.name}</span>
            ) : (
              <>
                <span>{donationsCount}</span> {donationsCountCaption}
              </>
            )}
          </div>
          <div>
            <span>{openAmount}</span> {openAmountCaption}
          </div>
        </div>
        <img src={orgaImageUrl} alt={project.carrier?.name} className="donatable-teaser--carrier-logo" />
      </div>

      <div className="donatable-teaser--progress">
        <div className="donatable-teaser--progress-bar" style={{ width: `${project.progress_percentage}%` }}></div>
      </div>

      {props.bottomContent}
    </a>
  )
}

ProjectTeaser.propTypes = {
  className: PropTypes.string,
  font: PropTypes.string,
  textColor: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
  progressbarColor: PropTypes.string,
  project: PropTypes.any.isRequired,
  showDescription: PropTypes.bool,
  extraHeight: PropTypes.number,
  showCarrierName: PropTypes.bool,
  bottomContent: PropTypes.element,
}
