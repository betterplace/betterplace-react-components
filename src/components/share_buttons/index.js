export * from './facebook_button'
export * from './twitter_button'
export * from './telegram_button'
export * from './facebook_messenger_button'
export * from './whatsapp_button'
export * from './email_button'
