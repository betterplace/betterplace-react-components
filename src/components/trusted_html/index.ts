export * from './trusted_html'
