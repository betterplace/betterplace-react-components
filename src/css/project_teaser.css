/* ---------------------------------- block --------------------------------- */

.project-teaser {
  cursor: pointer;
  background: white;
  font-family: "Fira Sans", Arial, sans-serif;
  border: 1px solid #ccc;
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  color: #000000;
  font-weight: 300;
  height: 324px;
  box-sizing: border-box;
  transition: 0.15s box-shadow ease-out;
  text-decoration: none;
}

.project-teaser:hover,
.project-teaser:focus {
  text-decoration: none;
  box-shadow: 0 0 10px 0 rgba(0,0,0,0.5);
  color: inherit;
}

.project-teaser:hover .project-teaser--title,
.project-teaser:focus .project-teaser--title {
  text-decoration: underline;
}

@media (max-width: 767px) {
  .project-teaser {
    width: 100%;
  }
}

/* ----------------------------- block modifier ----------------------------- */

.project-teaser--with-description {
  height: 411px;
}

.project-teaser--with-bottom-content {
  height: 374px;
}

.project-teaser--with-bottom-content .project-teaser--progress {
  border-radius: 0;
}

.project-teaser--with-bottom-content .project-teaser--progress-bar {
  border-radius: 0;
}

/* -------------------------------- elements -------------------------------- */

.project-teaser--profile-picture {
  height: 150px;
  background-size: cover;
  border-top-right-radius: 3px;
  border-top-left-radius: 3px;
}

.project-teaser--carrier-logo {
  width: 65px;
  height: 65px;
  margin-right: 13px;
  overflow: hidden;
}

.project-teaser--title {
  display: block;
  margin: 10px;
  font-size: 16px;
  line-height: 22px;
  height: 41px;
  font-weight: bold;
  overflow: hidden;
}

.project-teaser--description {
  position: relative;
  height: 71px;
  overflow: hidden;
  padding: 0 13px;
  font-size: 14px;
  line-height: 18px;
  color: #6a6a6a;
}

.project-teaser--description:after {
  content: '';
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  height: 18px;
  background: linear-gradient(180deg, rgba(255,255,255,0) 0%, rgba(255,255,255,1) 100%);
}

.project-teaser--divider {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  flex-grow: 1;
  margin-bottom: 10px;
}

.project-teaser--factlist {
  font-size: 14px;
  line-height: 22px;
  padding-left: 13px;
  width: calc(100% - 100px);
}

.project--teaser--factlist div {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.project--teaser--factlist span {
  font-weight: bold;
}

.project-teaser--progress {
  background-color: #AECFD1;
  height: 15px;
  width: 100%;
  overflow: hidden;
  border-bottom-right-radius: 3px;
  border-bottom-left-radius: 3px;
}

.project-teaser--progress-bar {
  background: #2B8475;
  transition: width 0.2s linear;
  height: 100%;
  overflow: hidden;
  border-bottom-right-radius: 3px;
  border-bottom-left-radius: 3px;
}
