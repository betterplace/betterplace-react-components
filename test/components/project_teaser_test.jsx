import { ProjectTeaser } from '../../src/components/project_teaser'

const project = {
  id: 1,
  title: 'Skateistan Afghanistan',
  carrier: {
    name: 'Skateistan',
    picture: {
      links: []
    }
  },
  description: 'this is amazing',
  donations_count: 1337,
  open_amount_in_cents: 4200,
  profile_picture: {
    links: []
  },
  links: [
    {
      "rel": "platform",
      "href": "bp.org/p123"
    }
  ]
}

describe('ProjectTeaser', () => {
  it('renders', () => {
    // languageGetter.mockReturnValue('en')
    const element = shallow(<ProjectTeaser project={project} locale='en' />)
    expect(element).toMatchSnapshot()
  })

  it('does not render the description by default', () => {
    const element = shallow(<ProjectTeaser project={project} locale='en' />)
    expect(element).toMatchSnapshot()
  })

  it('does render the description if showDescription is true', () => {
    const element = shallow(<ProjectTeaser project={project} showDescription={true} locale='en' />)
    expect(element).toMatchSnapshot()
  })

  it('uses a custom href if given', () => {
    const element = shallow(<ProjectTeaser project={project} href='coupons.org/p123' locale='en' />)
    expect(element).toMatchSnapshot()
  })

  it('is possible to force the locale', () => {
    const element = shallow(<ProjectTeaser project={project} href='coupons.org/p123' locale='en' />)
    expect(element).toMatchSnapshot()
  })
})
