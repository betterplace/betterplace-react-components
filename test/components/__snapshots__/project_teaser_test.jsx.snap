// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTeaser does not render the description by default 1`] = `
<styled.a
  href="bp.org/p123"
  onClick={[Function]}
  target="_parent"
  title="Skateistan Afghanistan"
>
  <styled.div
    alt="Skateistan Afghanistan"
  />
  <styled.h2
    dangerouslySetInnerHTML={
      Object {
        "__html": "Skateistan Afghanistan",
      }
    }
  />
  <styled.div>
    <styled.div>
      <div>
        <span>
          ,
        </span>
         
      </div>
      <div>
        <span />
         
        Spenden
      </div>
      <div>
        <span>
          NaN
        </span>
         
        fehlen noch
      </div>
    </styled.div>
    <styled.img
      alt="Skateistan"
    />
  </styled.div>
  <styled.div>
    <div
      className="bar"
    />
  </styled.div>
</styled.a>
`;

exports[`ProjectTeaser does render the description if showDescription is true 1`] = `
<styled.a
  href="bp.org/p123"
  onClick={[Function]}
  target="_parent"
  title="Skateistan Afghanistan"
>
  <styled.div
    alt="Skateistan Afghanistan"
  />
  <styled.h2
    dangerouslySetInnerHTML={
      Object {
        "__html": "Skateistan Afghanistan",
      }
    }
  />
  <styled.div
    dangerouslySetInnerHTML={
      Object {
        "__html": "this is amazing",
      }
    }
  />
  <styled.div>
    <styled.div>
      <div>
        <span>
          ,
        </span>
         
      </div>
      <div>
        <span />
         
        Spenden
      </div>
      <div>
        <span>
          NaN
        </span>
         
        fehlen noch
      </div>
    </styled.div>
    <styled.img
      alt="Skateistan"
    />
  </styled.div>
  <styled.div>
    <div
      className="bar"
    />
  </styled.div>
</styled.a>
`;

exports[`ProjectTeaser renders 1`] = `
<styled.a
  href="bp.org/p123"
  onClick={[Function]}
  target="_parent"
  title="Skateistan Afghanistan"
>
  <styled.div
    alt="Skateistan Afghanistan"
  />
  <styled.h2
    dangerouslySetInnerHTML={
      Object {
        "__html": "Skateistan Afghanistan",
      }
    }
  />
  <styled.div>
    <styled.div>
      <div>
        <span>
          ,
        </span>
         
      </div>
      <div>
        <span />
         
        Spenden
      </div>
      <div>
        <span>
          NaN
        </span>
         
        fehlen noch
      </div>
    </styled.div>
    <styled.img
      alt="Skateistan"
    />
  </styled.div>
  <styled.div>
    <div
      className="bar"
    />
  </styled.div>
</styled.a>
`;

exports[`ProjectTeaser uses a custom href if given 1`] = `
<styled.a
  href="coupons.org/p123"
  onClick={[Function]}
  target="_parent"
  title="Skateistan Afghanistan"
>
  <styled.div
    alt="Skateistan Afghanistan"
  />
  <styled.h2
    dangerouslySetInnerHTML={
      Object {
        "__html": "Skateistan Afghanistan",
      }
    }
  />
  <styled.div>
    <styled.div>
      <div>
        <span>
          ,
        </span>
         
      </div>
      <div>
        <span />
         
        Spenden
      </div>
      <div>
        <span>
          NaN
        </span>
         
        fehlen noch
      </div>
    </styled.div>
    <styled.img
      alt="Skateistan"
    />
  </styled.div>
  <styled.div>
    <div
      className="bar"
    />
  </styled.div>
</styled.a>
`;
